### PR TITLE
[WIP] Implement schema printing for NTuple

### PR DESCRIPTION
**Pull request by lesimon74**
This pull request adds a visitor infrastructure for the RField class.  A concrete implementation, the print visitor, is used to show information about the field names and types of an ntuple on the terminal.

### CODE REVIEW

File: tree/ntuple/v7/src/RFieldVisitor.cxx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+// spaces needed so that larger field indexes can be shown properly alligned.
+int ROOT::Experimental::FieldDistance(unsigned int x)
+{
+   return (
+      x < 10000
+         ? 0
+         : (x < 100000 ? 1 : (x < 1000000 ? 2 : (x < 10000000 ? 3 : (x < 100000000 ? 4 : (x < 1000000000 ? 5 : 10))))));
+}
+
+std::string ROOT::Experimental::CutIfNecessary(const std::string &toCut, unsigned int maxAvailableSpace)
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/src/RFieldVisitor.cxx]: I'm not a big fan of this name - maybe `CutStringAndAddEllipsisIfNeeded` - yes waaay longer, but it's rarely called, now clear that it's operating on a string. And it's a namespace-scope function so it must be very specific.

File: tree/ntuple/v7/src/RField.cxx [hunk 2]
```diff
@@ -159,6 +161,10 @@ void ROOT::Experimental::Detail::RFieldBase::ConnectColumns(RPageStorage *pageSt
    }
 }
 
+void ROOT::Experimental::Detail::RFieldBase::AcceptVisitor (RNTupleVisitor &fVisitor) const {
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/src/RField.cxx]: No space in front of `(` please.

File: tree/ntuple/v7/src/RNTuple.cxx [hunk 3]
```diff
@@ -101,6 +103,29 @@ std::string ROOT::Experimental::RNTupleReader::GetInfo(const ENTupleInfo what) {
    return "";
 }
 
+
+void ROOT::Experimental::RNTupleReader::Print(std::ostream &output, int width, unsigned int fNoFields) {
+    if(width < 30) {
+        std::cout << "The width is too small! Should be at least 30.\n";
+        return;
+    }
+    RPrintVisitor fPrintVisitor(output, width, fNoFields);
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/src/RNTuple.cxx]: `fPrintVisitor` is a local variable not a class member, should be called `printVisitor`.

File: tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+   virtual void visitField(const ROOT::Experimental::Detail::RFieldBase &fField) = 0;
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::RPrintVisitor
+\ingroup NTuple
+\brief Contains settings for printing and prints a summary of an RField instance.
+     
+ Instances of this class are currently only invoked by RNTupleReader::Print() -> RFieldBase::AcceptVisitor()
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx]: That doesn't match my expectation. I would gave expected that I can invoke `RNTuple::VisitFields(visitor)` which calls `visitor.VisitField()` for each of the fields in the ntuple? Maybe we should discuss that verbally :-)

File: tree/ntuple/v7/src/RNTuple.cxx [hunk 5]
```diff
@@ -101,6 +103,29 @@ std::string ROOT::Experimental::RNTupleReader::GetInfo(const ENTupleInfo what) {
...
+        output << '*'; output << " NTUPLE ";
+    for(int i = 0; i < (width/2 -4); ++i) output << '*'; output << '\n';
+    //CutIfNecessary defined in RFieldVisitor.hxx
+    output << "* Ntuple  : " << CutIfNecessary(GetName(),width-13) << std::setw(width-11-GetName().size()) << "*\n";
+    output << "* Entries : " << GetNEntries() << std::setw(width-11-NumDigits(GetNEntries())) << "*\n";
+    for(int i = 0; i < width; ++i)
+        output << "*"; output << '\n';
+    for(const auto& field: *(GetModel()->GetRootField())) {
+        output << "*";
+        field.AcceptVisitor(fPrintVisitor);
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/src/RNTuple.cxx]: Why this detour? Could this invoke `printVisitor.visitField(field)`?

File: tree/ntuple/v7/test/ntuple_print.cxx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            std::remove(fPath.c_str());
+        }
+    };
+    
+} // anonymous namespace
+
+TEST(RNtuplePrint, FullString)
+{
+    TFile *file = TFile::Open("test.root", "RECREATE");
+    FileRaii fileGuard("test.root");
```
Comment (Axel-Naumann) [inline tree/ntuple/v7/test/ntuple_print.cxx]: Could you maybe combine the RAII and the `TFile` construction? E.g. by adding a `FileRAII::GetFile()`? That way, `FileRAII` could close and delete, on all control paths.

**Comment by lesimon74** [comment]
The way to implement RNTupleReader::Print() is currently being reconsidered, so reviewing is not necessary at the moment.

File: tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+*/
+// clang-format on
+
+class RPrintVisitor : public RNTupleVisitor {
```
[the remark this answers is not in the document]
Comment (lesimon74) [reply tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx]: I believe it is better to see how things change and decide later if it should go in a separate header.

**Comment by jblomer** [comment]
Replaced by #4044

**Pull request closed without merging**